lib/cn/hblock_reader.c

```diff
@@ -0,0 +1,189 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+#define MTF_MOCK_IMPL_hblock_reader
+
+#include <assert.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <hse_util/compiler.h>
+#include <hse_util/event_counter.h>
+#include <hse_util/hse_err.h>
+#include <mpool/mpool_structs.h>
+#include <mpool/mpool.h>
+
+#include "hblock_reader.h"
+#include "kvs_mblk_desc.h"
+#include "omf.h"
+#include "wbt_reader.h"
+
+static bool HSE_NONNULL(1)
+hblock_hdr_valid(const struct hblock_hdr_omf *omf)
+{
+    const uint32_t version = omf_hbh_version(omf);
+    const uint32_t magic = omf_hbh_magic(omf);
+
+    return HSE_LIKELY(magic == HBLOCK_HDR_MAGIC && version == HBLOCK_HDR_VERSION);
+}
+
+static merr_t
+hbr_madvise_region(
+    struct kvs_mblk_desc *hblk_desc,
+    uint32_t pg,
+    const uint32_t pg_cnt,
+    const int advice)
+{
+    merr_t err = 0;
+    const uint32_t pg_max = pg + pg_cnt;
+
+    while (pg < pg_max) {
+        const uint32_t chunk = min_t(uint32_t, pg_max - pg, HSE_RA_PAGES_MAX);
+
+        err = mpool_mcache_madvise(
+            hblk_desc->map, hblk_desc->map_idx, PAGE_SIZE * pg, PAGE_SIZE * chunk, advice);
+        if (ev(err))
+            return err;
+
+        pg += chunk;
+    }
+
+    return err;
+}
+
+void
+hbr_madvise_kmd(struct kvs_mblk_desc *kblkdesc, struct wbt_desc *desc, int advice)
+{
+    merr_t err;
+    u32    pg = desc->wbd_first_page + desc->wbd_root + 1;
+    u32    pg_cnt = desc->wbd_kmd_pgc;
+
+    err = hbr_madvise_region(kblkdesc, pg, pg_cnt, advice);
```

Comment:
Change `hbr_madvise_region` to return `void` and rewrite this (and the leaf node version) as one liners:
```
hbr_madvise_region(esc->wbd_first_page + desc->wbd_root + 1, desc->wbd_kmd_pgc)
```

Reply:
I just copied the same pattern as kblock_builder.c I think. Can change it though

Reply:
Regarding, returning void, do you want me to just ignore the error from mpool_mcache_advise?

Reply:
You an `ev(1)` if there's an error which makes it visible in stats via REST.  That's about all we do now right?

Reply:
That's what I've been using ev() for..

Reply:
It seems so free to return errors up the stack to me. I think since this code is going to go away when ptombs are removed and it is the same as kblock_builder.c code, I don't feel inclined to change it unless someone feels so strongly the other way.

Reply:
It is not free with regard to error handling.  But there's no need to change it.  As you say: will likely go away soonish anyhow.

Reply:
Seems like a lot of our problem with error handling is that in some cases the only way a function can fail is if an invalid arg is passed and then we are left doing pointless `if (err)` checks the rest of the way up the stack when the code is only used internally. But since we control the function calls, the error will never occur. I advocated for stopping this in internal symbols, but others felt differently.